<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN">
<HTML
><HEAD
><TITLE
>CGAT User Manual</TITLE
><META
NAME="GENERATOR"
CONTENT="Modular DocBook HTML Stylesheet Version 1.76b+
"></HEAD
><BODY
CLASS="book"
BGCOLOR="#FFFFFF"
TEXT="#000000"
LINK="#0000FF"
VLINK="#840084"
ALINK="#0000FF"
><DIV
CLASS="BOOK"
><A
NAME="AEN1"
></A
><DIV
CLASS="TITLEPAGE"
><H1
CLASS="title"
><A
NAME="AEN3"
></A
>CGAT User Manual</H1
><H3
CLASS="author"
><A
NAME="AEN7"
></A
>Ikuo Uchiyama</H3
><DIV
CLASS="affiliation"
><SPAN
CLASS="orgname"
>&#13;			National Institute for Basic Biology,
			National Institutes of Natural Sciences
		<BR></SPAN
></DIV
><P
CLASS="copyright"
>Copyright &copy; 2006 by Ikuo Uchiyama</P
><HR></DIV
><DIV
CLASS="TOC"
><DL
><DT
><B
>Table of Contents</B
></DT
><DT
><A
HREF="#whatis"
>What is CGAT ?</A
></DT
><DT
><A
HREF="#install"
>Installing CGAT</A
></DT
><DD
><DL
><DT
><A
HREF="#AEN24"
>Requirements</A
></DT
><DT
><A
HREF="#AEN81"
>Download</A
></DT
><DT
><A
HREF="#AEN101"
>Install</A
></DT
><DD
><DL
><DT
><A
HREF="#AEN103"
>AlignmentViewer</A
></DT
><DT
><A
HREF="#AEN146"
>DataServer</A
></DT
></DL
></DD
></DL
></DD
><DT
><A
HREF="#alignmentviewer"
>The CGAT AlignmentViewer program</A
></DT
><DD
><DL
><DT
><A
HREF="#startAliView"
>Starting AlignmentViewer</A
></DT
><DD
><DL
><DT
><A
HREF="#loadData"
>Load data</A
></DT
></DL
></DD
><DT
><A
HREF="#main"
>The main window of AlignmentViewer</A
></DT
><DD
><DL
><DT
><A
HREF="#control_panel"
>Control panel</A
></DT
><DT
><A
HREF="#alignmentDisplay"
>Alignment display panel</A
></DT
><DT
><A
HREF="#dotplot"
>Dotplot display panel</A
></DT
></DL
></DD
><DT
><A
HREF="#MenuItems"
>Menu items</A
></DT
><DD
><DL
><DT
><A
HREF="#File"
>File menu</A
></DT
><DT
><A
HREF="#Search"
>Search menu</A
></DT
><DT
><A
HREF="#View"
>View menu</A
></DT
></DL
></DD
><DT
><A
HREF="#configAlignViewer"
>Configuring AlignmentViewer</A
></DT
></DL
></DD
><DT
><A
HREF="#dataserver"
>The CGAT DataServer package</A
></DT
><DD
><DL
><DT
><A
HREF="#overview"
>Overview of CGAT DataServer</A
></DT
><DD
><DL
><DT
><A
HREF="#directory"
>Directory structure</A
></DT
></DL
></DD
><DT
><A
HREF="#AEN678"
>Database construction</A
></DT
><DD
><DL
><DT
><A
HREF="#databuild"
>Overview of the data construction procedure</A
></DT
><DT
><A
HREF="#prepareData"
>Preparing genomic sequences</A
></DT
><DT
><A
HREF="#configBuild"
>Configuring database building procedure</A
></DT
><DT
><A
HREF="#startBuild"
>Start building</A
></DT
><DT
><A
HREF="#releaseData"
>Release data</A
></DT
></DL
></DD
><DT
><A
HREF="#AEN796"
>Web server set up</A
></DT
></DL
></DD
></DL
></DIV
><DIV
CLASS="chapter"
><HR><H1
><A
NAME="whatis"
></A
> What is CGAT ? </H1
><P
>&#13;<TT
CLASS="application"
>CGAT</TT
> is a Comparative Genome Analysis Tool
that was developed for
detailed comparison of closely related bacterial-sized genomes, mainly for
analysis of complex evolutionary changes between two related
genomes.  CGAT visualizes precomputed pairwise genome alignments
on both dotplot and alignment viewers.  Users can put several
information on this alignment, such as existence of tandem repeats
or interspersed repetitive sequences and changes in codon usage bias,
to facilitate interpretation of the observed
genomic changes.  Besides visualization functionalities,
CGAT also provides a general framework to process genome-scale
alignments using various existent alignment programs such as
BLAST/MegaBlast, FASTA, MUMmer, Blat and BlastZ.
    </P
><P
>&#13;Basically, CGAT employs a client-server architecture, which consists of
<B
CLASS="command"
>AlignmentViewer</B
> (client; a Java application) and
<B
CLASS="command"
>DataServer</B
>
(a set of Perl scripts).
The DataServer consists of data construction scripts and CGI scripts
and AlignmentViewer obtains the data from the server thorough the HTTP
protocol. AlignmentViewer can also obtain data from the local file system
when the server and client are installed on the same machine.
    </P
></DIV
><DIV
CLASS="chapter"
><HR><H1
><A
NAME="install"
></A
> Installing CGAT </H1
><DIV
CLASS="sect1"
><H2
CLASS="sect1"
><A
NAME="AEN24"
></A
>Requirements</H2
><P
>&#13;<TT
CLASS="application"
>AlignmentViewer</TT
> is written in Java and
should run on any machine supporting
<A
HREF="http://www.java.com"
TARGET="_top"
>Java</A
> (JDK 1.4 or higher).
      </P
><P
>&#13;<TT
CLASS="application"
>DataServer</TT
> is a set of Perl scripts
and assumes Unix environment with <A
HREF="http://www.perl.com"
TARGET="_top"
>Perl</A
>
installed (version 5.8 or higher). It also assumes that the perl executable
is installed in the <TT
CLASS="filename"
>/usr/bin</TT
> directory; please make
a symbolic link if it is located in a different directory.
The LWP (libwww-perl) module, which is downloadable from
<A
HREF="http://www.cpan.org"
TARGET="_top"
>the CPAN site</A
>,
is also required when you want to download data from the MBGD server.
DataServer also needs an HTTPD server for communicating with
the client program. In this manual, we assume that
<A
HREF="http://httpd.apache.org"
TARGET="_top"
>Apache HTTPD</A
> version 2.0 or
more is installed on your server machine.
      </P
><P
>&#13;CGAT uses several programs for calculating alignments of two sequences
and identifying several feature segments on individual sequences.
Currently the following programs are supported:
      </P
><DIV
CLASS="table"
><A
NAME="AEN36"
></A
><P
><B
>Table 1.  Alignment programs currently supported by CGAT </B
></P
><TABLE
BORDER="1"
BGCOLOR="#E0E0E0"
CELLSPACING="0"
CELLPADDING="4"
CLASS="CALSTABLE"
><THEAD
><TR
><TH
ALIGN="LEFT"
VALIGN="MIDDLE"
>program</TH
><TH
ALIGN="LEFT"
VALIGN="MIDDLE"
>URL</TH
></TR
></THEAD
><TBODY
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> Blast/MegaBlast </TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> <A
HREF="http://www.ncbi.nlm.nih.gov/BLAST/"
TARGET="_top"
>http://www.ncbi.nlm.nih.gov/BLAST/</A
> </TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> Fasta</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> <A
HREF="http://fasta.bioch.virginia.edu/"
TARGET="_top"
>http://fasta.bioch.virginia.edu/</A
> </TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> MUMmer</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> <A
HREF="http://mummer.sourceforge.net/"
TARGET="_top"
>http://mummer.sourceforge.net/</A
> </TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> Blat/WABA </TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> <A
HREF="http://www.cse.ucsc.edu/~kent"
TARGET="_top"
>http://www.cse.ucsc.edu/~kent</A
> </TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> BlastZ</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> <A
HREF="http://www.bx.psu.edu/miller_lab/"
TARGET="_top"
>http://www.bx.psu.edu/miller_lab/</A
> </TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> PatternHunter </TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> <A
HREF="http://www.bioinformaticssolutions.com/products/ph/"
TARGET="_top"
>http://www.bioinformaticssolutions.com/products/ph/</A
> </TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> CHAOS </TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> <A
HREF="http://lagan.stanford.edu/lagan_web/index.shtml"
TARGET="_top"
>http://lagan.stanford.edu/lagan_web/index.shtml</A
> </TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> GAME </TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> <A
HREF="http://cancer.informatics.indiana.edu/game/"
TARGET="_top"
>http://cancer.informatics.indiana.edu/game/</A
> </TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> SSAHA </TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> <A
HREF="http://www.sanger.ac.uk/Software/analysis/SSAHA/"
TARGET="_top"
>http://www.sanger.ac.uk/Software/analysis/SSAHA/</A
> </TD
></TR
></TBODY
></TABLE
></DIV
><P
>&#13;You should install one or more of the above programs on the server machine
before installing the CGAT DataServer.  Please refer to the above URL for
installing individual programs.  Basically, the DataServer assumes
at least the NCBI Blast/MegaBlast package is installed.
      </P
></DIV
><DIV
CLASS="sect1"
><HR><H2
CLASS="sect1"
><A
NAME="AEN81"
></A
>Download</H2
><P
>&#13;The latest version of CGAT is available at
<A
HREF="http://mbgd.genome.ad.jp/CGAT/"
TARGET="_top"
>http://mbgd.genome.ad.jp/CGAT/</A
>.

You can download and install only the client program
(<TT
CLASS="filename"
>AlignmentViewer.jar</TT
>)
or the full version including both the client and server programs
(<TT
CLASS="filename"
>CGAT.tar</TT
>) depending on what you need.
       </P
><P
></P
><UL
><LI
><P
>&#13;The client program (
<A
HREF="http://mbgd.genome.ad.jp/CGAT/dist/AlignmentViewer.jar"
TARGET="_top"
>&#13;<TT
CLASS="filename"
>AlignmentViewer.jar</TT
></A
>).
You can download and install only the client program
if you intend to connect to the existing servers that you have
already set up or to the default server
(<A
HREF="http://mbgd.genome.ad.jp/CGAT/"
TARGET="_top"
>http://mbgd.genome.ad.jp/CGAT/</A
>) for trial.
      </P
></LI
><LI
><P
>&#13;The client program as a Mac OSX application (
<A
HREF="http://mbgd.genome.ad.jp/CGAT/dist/AlignmentViewer.app.tgz"
TARGET="_top"
>&#13;<TT
CLASS="filename"
>AlignmentViewer.app</TT
></A
>).
We recommend Mac OSX users to download
this because it has been set up with increased memory
allocation that solves the problem described in the next section.
      </P
></LI
><LI
><P
>&#13;The full version of CGAT (
<A
HREF="http://mbgd.genome.ad.jp/CGAT/dist/CGAT.tgz"
TARGET="_top"
>&#13;<TT
CLASS="filename"
>CGAT.tgz</TT
></A
>),
including both the client and server programs.
You must install this package if you intend to analyze your own data locally.
      </P
></LI
></UL
></DIV
><DIV
CLASS="sect1"
><HR><H2
CLASS="sect1"
><A
NAME="AEN101"
></A
>Install</H2
><DIV
CLASS="sect2"
><H3
CLASS="sect2"
><A
NAME="AEN103"
></A
>AlignmentViewer</H3
><P
>&#13;	Basically, you need not particular operation after having downloaded
	the program file (<TT
CLASS="filename"
>AlignmentViewer.jar</TT
>) &#8212;
	double click the icon (OSX or Windows) or type the java command
	(Unix) to invoke the program. See the chapter
	<SPAN
CLASS="QUOTE"
>"
	<A
HREF="#alignmentviewer"
><I
>The CGAT AlignmentViewer program</I
></A
>
	"</SPAN
>
      </P
><P
>&#13;	However, in the default setting of the Java virtual machine (JVM),
	you will be likely to encounter a memory allocation error
	especially when aligning long sequences dynamically.
	To avoid this problem, you need to increase the size of
	the memory allocation pool for the JVM.
      </P
><P
></P
><UL
><LI
><P
>&#13;	<SPAN
CLASS="productname"
>Windows</SPAN
>:
	Download the <TT
CLASS="filename"
>AlignmentViewer.jar</TT
> file and
	do the following (see <A
HREF="#fig_win_shortcut"
>Figure 1</A
>).
      <P
></P
><OL
TYPE="1"
><LI
><P
>&#13;	Create a shortcut of the <TT
CLASS="filename"
>AlignmentViewer.jar</TT
>
	file (Right-click on the jar file &#8658;
	<B
CLASS="guimenu"
>Create Shortcut</B
>).
	</P
></LI
><LI
><P
>&#13;	Open the properties window of the shortcut (Right-click on the
	shortcut &#8658; <B
CLASS="guimenu"
>Properties</B
>).
	</P
></LI
><LI
><P
>&#13;	Open the Shortcut tab and
	enter the java command in the <B
CLASS="guimenu"
>Target:</B
> box as
	follows:
	[ <TT
CLASS="userinput"
><B
>java -Xmx512m -jar "C:\Documents and Settings\..."</B
></TT
> ],
	where <TT
CLASS="userinput"
><B
>"C:\Documents and Settings\..."</B
></TT
> is the
	original file location. This might be different in you situation,
	but do not modify the original name in anyway. Then, press "OK".
	</P
></LI
></OL
>
      </P
></LI
><LI
><P
>&#13;	<SPAN
CLASS="productname"
>Mac OSX</SPAN
>:
	Download and use the OSX application version of AlignmentViewer
	(<TT
CLASS="filename"
>AlignmentViewer.app</TT
>),
	instead of the <TT
CLASS="filename"
>AlignmentViewer.jar</TT
> file.
      </P
></LI
><LI
><P
>&#13;	<SPAN
CLASS="productname"
>Linux/Unix</SPAN
>:
	Download the <TT
CLASS="filename"
>AlignmentViewer.jar</TT
> file and
	type <TT
CLASS="userinput"
><B
>java -Xmx512m -jar AlignmentViewer.jar</B
></TT
>
	to invoke the program.
      </P
></LI
></UL
><DIV
CLASS="figure"
><A
NAME="fig_win_shortcut"
></A
><DIV
CLASS="mediaobject"
><P
><IMG
SRC="fig/win_shortcut.gif"></P
></DIV
><P
><B
>Figure 1. Increasing the size of the memory allocation pool of the JVM
for AlignmentViewer on Windows XP</B
></P
></DIV
></DIV
><DIV
CLASS="sect2"
><HR><H3
CLASS="sect2"
><A
NAME="AEN146"
></A
>DataServer</H3
><P
>&#13;Copy the downloaded archive <TT
CLASS="filename"
>CGAT.tgz</TT
> into
an appropriate directory and extract files by <B
CLASS="command"
>tar</B
> command,
which generates a directory named <TT
CLASS="filename"
>CGAT</TT
>.
Enter this directory and invoke a configuration script
<B
CLASS="command"
>SetUp.pl</B
>.
<TABLE
BORDER="0"
BGCOLOR="#E0E0E0"
WIDTH="100%"
><TR
><TD
><PRE
CLASS="screen"
>&#13;tar cvfz CGAT.tgz
cd CGAT
SetUp.pl
</PRE
></TD
></TR
></TABLE
>
SetUp.pl first asks you to set the <TT
CLASS="envar"
>CGAT_HOME</TT
>
environmental variable to the root of the generated directory tree
(hereafter we call this directory <TT
CLASS="envar"
>$CGAT_HOME</TT
>), 
and subsequently it sets several command paths and compiles some programs.
SetUp.pl saves the <TT
CLASS="envar"
>CGAT_HOME</TT
> path in
scripts <TT
CLASS="filename"
>etc/cgat.sh</TT
> for sh and bash
and <TT
CLASS="filename"
>etc/cgat.csh</TT
> for csh and tcsh.
Before building the database or doing other operation, you must source the
above script.

<TABLE
BORDER="0"
BGCOLOR="#E0E0E0"
WIDTH="100%"
><TR
><TD
><PRE
CLASS="screen"
>&#13;# in csh or tcsh
source etc/cgat.csh

# in sh or bash
. etc/cgat.sh
</PRE
></TD
></TR
></TABLE
>
We recommend you to add this command to your .cshrc script
(or .profile, .bashrc etc.)
(in this case, you must use the absolute path
rather than the relative path, of course).
	</P
><P
>&#13;The set up script tries to detect alignment programs available
on the server machine and installs detected programs
in the <TT
CLASS="filename"
>$CGAT_HOME/build/align</TT
> directory.
You might want to move some commands from the <TT
CLASS="filename"
>&#13;$CGAT_HOME/build/align.in</TT
> directory
or to remove some unnecessary commands manually.
	</P
><P
>&#13;Similarly, you might want to install manually some optional commands for
collecting feature segments from the <TT
CLASS="filename"
>$CGAT_HOME/build/segment.opt
</TT
> directory. Note that in this case you must always set up
by yourself (if you wanted), since currently the set up script does not
try to check the availability of these commands automatically.
Currently, the following commands are available:
	</P
><P
></P
><UL
><LI
><P
>&#13;<B
CLASS="command"
>Glimmer</B
>:
Microbial gene-finding system available at
<A
HREF="http://www.cbcb.umd.edu/software/glimmer/"
TARGET="_top"
>http://www.cbcb.umd.edu/software/glimmer/</A
>.
Currently, ver.2 is assumed by default, but you can use ver.3
if you manually edit the <TT
CLASS="filename"
>commands/Glimmer.pl</TT
> file.
Before that, of course, you must correctly set up the glimmer program itself.
	</P
></LI
><LI
><P
>&#13;<B
CLASS="command"
>ISfind</B
>:
Searching an insertion sequence (IS) database by the BLAST program.
The IS database should be a FASTA formatted sequence file with
each comment line containing IS family name as
<TT
CLASS="computeroutput"
>"&#62;<TT
CLASS="parameter"
><I
>${ISname}</I
></TT
> 
Fam:<TT
CLASS="parameter"
><I
>${familyname}</I
></TT
>"</TT
>,
and the file should be installed as <TT
CLASS="filename"
>$CGAT_HOME/database/isdb
</TT
>. See <A
HREF="http://www-is.biotoul.fr"
TARGET="_top"
>the IS Finder
database</A
> for information about IS families. One can use the
<A
HREF="http://bioportal.ddbj.nig.ac.jp/is/"
TARGET="_top"
>GIB-IS</A
> database
available at DDBJ as a downloadable IS database. Download the table file
(<TT
CLASS="filename"
>IS-table-text.txt</TT
>) and run the following command:
<TABLE
BORDER="0"
BGCOLOR="#E0E0E0"
WIDTH="100%"
><TR
><TD
><PRE
CLASS="screen"
>&#13;$CGAT_HOME/build/bin/conv_GIB_IS.pl -nr IS-table-text.txt &#62; $CGAT_HOME/database/isdb
</PRE
></TD
></TR
></TABLE
>
	</P
></LI
></UL
></DIV
></DIV
></DIV
><DIV
CLASS="chapter"
><HR><H1
><A
NAME="alignmentviewer"
></A
> The CGAT AlignmentViewer program</H1
><DIV
CLASS="sect1"
><H2
CLASS="sect1"
><A
NAME="startAliView"
></A
>Starting AlignmentViewer</H2
><P
>&#13;	On Windows or Mac OSX, double clicking on the "AlignmentViewer"
	icon invokes the program.
	On Unix systems, type the following command on the command line:
<TABLE
BORDER="0"
BGCOLOR="#E0E0E0"
WIDTH="100%"
><TR
><TD
><PRE
CLASS="screen"
>&#13;	java -jar AlignmentViewer.jar
</PRE
></TD
></TR
></TABLE
>
	</P
><P
>&#13;	However, when invoking with the above command, you will be
	likely to encounter a memory allocation error especially when you
	want to see an alignment of long sequences generated by dynamical
	calculation.
	To avoid this, we recommend you to increase the size of the memory
	allocation pool for the Java virtual machine with
	the <TT
CLASS="option"
>-Xmx</TT
> option.
	</P
><TABLE
BORDER="0"
BGCOLOR="#E0E0E0"
WIDTH="100%"
><TR
><TD
><PRE
CLASS="screen"
>&#13;	java -Xmx512m -jar AlignmentViewer.jar
</PRE
></TD
></TR
></TABLE
><DIV
CLASS="sect2"
><HR><H3
CLASS="sect2"
><A
NAME="loadData"
></A
>Load data</H3
><P
>&#13;	After running the program, first you must load data by choosing
	<B
CLASS="guimenu"
>File &#8658; Load</B
> from the menu.
	</P
><DIV
CLASS="figure"
><A
NAME="AEN195"
></A
><DIV
CLASS="mediaobject"
><P
><IMG
SRC="fig/dataload.gif"></P
></DIV
><P
><B
>Figure 1. The menu window for specifying data to be loaded</B
></P
></DIV
><P
></P
><DIV
CLASS="variablelist"
><DL
><DT
>Server URL</DT
><DD
><P
>&#13;Choose the server from which data are downloaded. By default, only
the default CGAT server (running on the MBGD server) is registered.
You can add more servers to this list in the Locations section
of the Properties panel (<B
CLASS="guimenu"
>File &#8658; Properties</B
>),
if you or your colleagues have been setup an original new server.
See the chapter <SPAN
CLASS="QUOTE"
>"
<A
HREF="#dataserver"
><I
>The CGAT DataServer package</I
></A
>
"</SPAN
> for server setup and
the <SPAN
CLASS="QUOTE"
>"
<A
HREF="#configAlignViewer"
><I
>Configuring AlignmentViewer</I
></A
>
"</SPAN
>section for properties setup.
</P
></DD
><DT
>Alignment Data</DT
><DD
><P
>&#13;Choose the alignment data to be downloaded specified by the combination of
organism names and a program name.
Note that choosing the server URL above reset the list of
available alignment data.
If you have setup the data on the local file system that the AlignmentViewer
program can directly accessed, and have correctly set the
<TT
CLASS="envar"
>CGAT_HOME</TT
> variable in the Properties panel,
you can also choose the data available
on the local machine. These data are listed without the "(Server)" indication.
</P
></DD
><DT
>Gene Attribute</DT
><DD
><P
>&#13;Choose the gene attribute data for coloring genes.
</P
></DD
><DT
>Feature Segment Data</DT
><DD
><P
>&#13;Choose the feature segment data to be shown on the additional annotation
tracks (see the section
<SPAN
CLASS="QUOTE"
>"<A
HREF="#alignmentDisplay"
><I
>Alignment display panel</I
></A
>
"</SPAN
>). By default, maximally three types of segment data can
be loaded at once, but the maximum number can be increased in
the Properties panel.
</P
></DD
></DL
></DIV
></DIV
></DIV
><DIV
CLASS="sect1"
><HR><H2
CLASS="sect1"
><A
NAME="main"
></A
>The main window of AlignmentViewer</H2
><P
>&#13;The main window of AlignmentViewer consists of four parts: a menu bar,
a control panel, an alignment display panel, and a dotplot display panel.
Both the alignment and dotplot displays show precomputed alignments.
Users can change the current view on each display by pressing a scrolling
or zooming button on the control panel. These operations update both
the alignment and dotplot displays in a coherent manner.
	</P
><P
>&#13;In AlignmentViewer, one of the compared sequences is considered as
the reference sequence and the other is considered as the
target sequence.
In the alignment display panel, the reference and the target sequences
are displayed on the upper and lower tracks, respectively,
and in the dotplot display panel, the reference and the target
sequences  correspond to the x- and y-axes, respectively.
These sequences can be exchanged by <B
CLASS="guimenu"
>View</B
>
&#8658; <B
CLASS="guimenu"
>Exchange Reference Genome</B
>.

	</P
><DIV
CLASS="figure"
><A
NAME="AEN233"
></A
><DIV
CLASS="mediaobject"
><P
><IMG
SRC="fig/mainwindow.gif"></P
></DIV
><P
><B
>Figure 2.  Main window of AlignmentViewer</B
></P
></DIV
><DIV
CLASS="sect2"
><HR><H3
CLASS="sect2"
><A
NAME="control_panel"
></A
>Control panel</H3
><P
>&#13;	The control panel contains several buttons for controlling the current
	view of both the alignment and dotplot displays.

<DIV
CLASS="figure"
><A
NAME="AEN243"
></A
><DIV
CLASS="mediaobject"
><P
><IMG
SRC="fig/controlpanel.gif"></P
></DIV
><P
><B
>Figure 3.  Control panel </B
></P
></DIV
>

<P
></P
><DIV
CLASS="variablelist"
><DL
><DT
>Move left/right</DT
><DD
><P
>&#13;Move the current position to the left or right along the current alignment.
</P
></DD
><DT
>Zoom in/out</DT
><DD
><P
>&#13;Zoom-in or -out the current view.
</P
></DD
><DT
>Alignment color</DT
><DD
><P
>&#13;Toggle the alignment color mode between the "Best Hit" mode and
the "Identity" mode.
	<P
></P
><UL
><LI
><P
>&#13;	Best Hit: white, orthologs; cyan, duplicated on the sequence 1;
	green, duplicated on the sequence 2; and blue, paralogs (by default).
	</P
></LI
><LI
><P
>&#13;	Identity: red, 50%; yellow, 85%; and green, 100% (by default).
	</P
></LI
></UL
>
</P
></DD
><DT
>Alignment mode</DT
><DD
><P
>&#13;Toggle the alignment mode between the "Regionwise" alignment mode and
the "Reference-target" alignment mode.
See the section <SPAN
CLASS="QUOTE"
>"
<A
HREF="#alignmentDisplay"
><I
>Alignment display panel</I
></A
> "</SPAN
>.
</P
></DD
><DT
>Exchange ref-target</DT
><DD
><P
>&#13;Exchange the reference sequence and the target sequence.
</P
></DD
><DT
>Display dotplot</DT
><DD
><P
>&#13;Toggle the status of the dotplot display panel between displayed and hidden.
</P
></DD
><DT
>Zoom in/out dotplot</DT
><DD
><P
>&#13;Zoom-in or -out the dotplot display.
</P
></DD
><DT
>Zoomed axis</DT
><DD
><P
>&#13;Specify the axis of the dotplot to be zoomed individually.
</P
></DD
><DT
>Change background</DT
><DD
><P
>&#13;Toggle the background color of the alignment and dotplot displays
between the dark color (black by default) and the light color
(white by default).
</P
></DD
><DT
>Show annotation table</DT
><DD
><P
>&#13;Show an annotation table for information on gene or feature segment.
</P
></DD
></DL
></DIV
>
	</P
><DIV
CLASS="figure"
><A
NAME="AEN298"
></A
><DIV
CLASS="mediaobject"
><P
><IMG
SRC="fig/alignment_color.gif"></P
></DIV
><P
><B
>Figure 4. Alignment color modes. The Best-Hit mode (A),
and the Identity mode (B)</B
></P
></DIV
></DIV
><DIV
CLASS="sect2"
><HR><H3
CLASS="sect2"
><A
NAME="alignmentDisplay"
></A
>Alignment display panel</H3
><P
>&#13;The alignment display panel contains three basic tracks: the central
tracks (alignment track) shows a graphical representation of the
alignments, and the upper and lower tracks (gene annotation tracks) show
the gene annotation of the upper and lower genomes, respectively.
More annotation tracks can be added to display the location of
several feature segments.
	</P
><P
>&#13;In addition, the "location fields" and the "strand switches" are on the left
side of the tracks.
The location field indicates the central coordinate of the current view,
and strand switch indicates the strand of each genome currently selected.
Users can jump to the desired location by entering a coordinate value
or a gene identifier (e.g. HP0812) in the location field, and can toggle
the strand direction by pressing the strand switch.
	</P
><P
>&#13;Some notes: 1) You must always enter the location of the reference genome
if you want to see the orthologous alignment around the specified locus.
If you want to specify the location in the current target genome,
you should exchanges the reference and the target genomes before
entering the position; 2) If you invert the reference genome by the
strand switch button, the direction of the subsequent movement is also
inverted. This means that when you press the 'move left' button,
the cursor in the dotplot display will move right, since the coordinate
of the dotplot display is not changed by this operation.
	</P
><DIV
CLASS="figure"
><A
NAME="AEN310"
></A
><DIV
CLASS="mediaobject"
><P
><IMG
SRC="fig/alignment_display.gif"></P
></DIV
><P
><B
>Figure 5.  Alignment display panel</B
></P
></DIV
><DIV
CLASS="sect3"
><HR><H4
CLASS="sect3"
><A
NAME="AEN317"
></A
>Alignment modes</H4
><P
>&#13;In CGAT, there are two modes in the alignment display:
"regionwise" alignment mode and "reference-target" alignment mode.

In the regionwise alignment mode, all alignments included in the current view
are displayed. Each alignment is represented as a trapezoid formed by
connecting the corresponding endpoints of the aligned region.

On the other hand, in the reference-target alignment mode, every
orthologous regions of the target sequence is mappend as a fragment
onto the reference sequence.
	</P
><DIV
CLASS="figure"
><A
NAME="AEN320"
></A
><DIV
CLASS="mediaobject"
><P
><IMG
SRC="fig/alignment_mode.gif"></P
></DIV
><P
><B
>Figure 6.  The two alignment modes: the regionwise mode (A) and
the reference-target mode (B). </B
></P
></DIV
></DIV
><DIV
CLASS="sect3"
><HR><H4
CLASS="sect3"
><A
NAME="AEN327"
></A
>Alignment track</H4
><P
>&#13;The alignment track displays precomputed alignments schematically.
Using the zooming buttons in the control panel, users can change
the scale from the entire genome level to the single nucleotide
level. In the nucleotide sequence alignment, mismatched site is colored
red and gapped site is colored green (by default).
	</P
><DIV
CLASS="figure"
><A
NAME="AEN330"
></A
><DIV
CLASS="mediaobject"
><P
><IMG
SRC="fig/zoom_alignment.gif"></P
></DIV
><P
><B
>Figure 7.  Zooming the alignment display </B
></P
></DIV
><P
>&#13;In the regionwise alignment mode, left click on one of the alignment
regions (tetragons) sets the center of the selected alignment adjusted to
the center of the display.
	</P
><P
>&#13;By right click on the alignment track, a pop-up menu appears including
the following items:
<P
></P
><UL
><LI
><P
> <B
CLASS="guimenu"
>View Re-Alignment</B
>:
	Re-calculate alignment between the sequences within the current
	displayed region. This item appears only when the current
	displayed region is smaller than the specified length
	(Size of Realignment).
	See the <SPAN
CLASS="QUOTE"
>"<A
HREF="#configAlignViewer"
><I
>Configuring AlignmentViewer</I
></A
>"</SPAN
> section.
</P
></LI
><LI
><P
> <B
CLASS="guimenu"
>View Alignment</B
>:
	Display the precomputed alignment under the clicked position.
</P
></LI
><LI
><P
> <B
CLASS="guimenu"
>Ident = ## %</B
>:
	No action. The item is displayed only for showing the percent
	identity of the precomputed alignment under the clicked position.
</P
></LI
><LI
><P
> <B
CLASS="guimenu"
>Color Legend</B
>:
	Display the color legend for the current alignment display.
</P
></LI
></UL
>
	</P
></DIV
><DIV
CLASS="sect3"
><HR><H4
CLASS="sect3"
><A
NAME="AEN354"
></A
>Annotation track</H4
><P
>&#13;The annotation track displays several types of feature segments
identified in each genome. 
	</P
><P
>&#13;By left click on one of the feature segments in the annotation track,
a pop-up menu appears:
<P
></P
><UL
><LI
><P
> <TT
CLASS="varname"
>SegmentName</TT
>:
	Open the web browser to access the web site,
	if an appropriate URL is defined on the server.
	Otherwise, no action is taken.
</P
></LI
><LI
><P
> <B
CLASS="guimenu"
>View Sequence: ## - ##</B
>:
	Show the nucleotide sequence of the feature segment.
</P
></LI
></UL
>

By right click on the annotation track, another pop-up menu appears:

<P
></P
><UL
><LI
><P
> <B
CLASS="guimenu"
>Clicked Pos: ## </B
>:
	No action. The item is displayed only for showing the clicked position.
</P
></LI
><LI
><P
> <B
CLASS="guimenu"
>Redraw (Set Center=##) </B
>:
	Move the clicked position to the center of the display window.
</P
></LI
><LI
><P
> <B
CLASS="guimenu"
>Search Alignment ## - ## </B
>:
	Search the set of precomputed alignments for those covering
	the region around the clicked position.
</P
></LI
><LI
><P
> <B
CLASS="guimenu"
>Color Legend</B
>:
	Display the color legend for the current annotation display.
</P
></LI
></UL
>

	</P
></DIV
></DIV
><DIV
CLASS="sect2"
><HR><H3
CLASS="sect2"
><A
NAME="dotplot"
></A
>Dotplot display panel</H3
><P
>&#13;In the dotplot display panel, all alignments are plotted onto
the 2-dimensional space with the horizontal (X) and the vertical (Y)
axes representing the reference and the target sequences, respectively.
In this plot, the current scope, i.e. the region displayed in the alignment
display, is also shown as a rectangular region.
Users can move the current scope by clicking on any points in the plot
to move.
	</P
><DIV
CLASS="figure"
><A
NAME="AEN381"
></A
><DIV
CLASS="mediaobject"
><P
><IMG
SRC="fig/dotplot_display.gif"></P
></DIV
><P
><B
>Figure 8.  Dotplot display panel </B
></P
></DIV
><P
>&#13;In the dotplot display, independent scaling of the x-axis and y-axis
is possible. This feature is useful in visualizing the distribution
of homologous regions of a specific segment on one genome against
the entirety of the other genome.
In the following example, dotplot with expanded y-axis scale
clarifies the presence of inverted duplication at both ends
of the rearranged segment on the reference genome (x-axis).
	</P
><DIV
CLASS="figure"
><A
NAME="AEN389"
></A
><DIV
CLASS="mediaobject"
><P
><IMG
SRC="fig/dotplot_zoom_y.gif"></P
></DIV
><P
><B
>Figure 9.  Dotplot display comparing the entire genomes (A) and
	dotplot display with expanded y-axis scale</B
></P
></DIV
></DIV
></DIV
><DIV
CLASS="sect1"
><HR><H2
CLASS="sect1"
><A
NAME="MenuItems"
></A
>Menu items</H2
><DIV
CLASS="sect2"
><H3
CLASS="sect2"
><A
NAME="File"
></A
>File menu</H3
><DIV
CLASS="sect3"
><H4
CLASS="sect3"
><A
NAME="Load"
></A
>Load</H4
><P
>&#13;Load data from the specified server or the local file system.
See the <SPAN
CLASS="QUOTE"
>"<A
HREF="#startAliView"
><I
>Starting AlignmentViewer</I
></A
>
"</SPAN
> section.
	  </P
></DIV
><DIV
CLASS="sect3"
><HR><H4
CLASS="sect3"
><A
NAME="Print"
></A
>Print</H4
><P
>&#13;Print out the image of the current alignment display.
	  </P
></DIV
><DIV
CLASS="sect3"
><HR><H4
CLASS="sect3"
><A
NAME="Properties"
></A
>Properties</H4
><P
>&#13;	Configure user property parameters.
	See the <SPAN
CLASS="QUOTE"
>"<A
HREF="#configAlignViewer"
><I
>Configuring AlignmentViewer</I
></A
> "</SPAN
> section.
	</P
></DIV
><DIV
CLASS="sect3"
><HR><H4
CLASS="sect3"
><A
NAME="Exit"
></A
>Exit</H4
><P
>&#13;Exit the program.
This item appears only in the standalone application.
For applet version, you must exit the web browser to exit the program.
	  </P
></DIV
></DIV
><DIV
CLASS="sect2"
><HR><H3
CLASS="sect2"
><A
NAME="Search"
></A
>Search menu</H3
><DIV
CLASS="sect3"
><H4
CLASS="sect3"
><A
NAME="MarkGenes"
></A
>Mark genes</H4
><P
>&#13;Mark a specified set of genes.
	  </P
><P
>&#13;An input form is opened for users to enter a list of gene names to be
marked. Gene names can be entered individually or be read from a file.
By pressing the <B
CLASS="guibutton"
>Search</B
> button, gene names are
registered.  Locations of the registered genes are indicated
on each axis of the dotplot display, and a new table window is
opened containing the information of the these genes.
	  </P
><P
>&#13;Additional calls of this function will add genes to the existing list.
To remove genes from the list, enter the gene names to be removed and
press <B
CLASS="guibutton"
>Clear</B
>.
To clear the entire list,  press <B
CLASS="guibutton"
>All Clear</B
>.
	  </P
><DIV
CLASS="figure"
><A
NAME="AEN426"
></A
><DIV
CLASS="mediaobject"
><P
><IMG
SRC="fig/mark_genes.gif"></P
></DIV
><P
><B
>Figure 10.  Marking genes. Input form for entering genes (A) and the dotplot
with genes marked with triangles on each axis (B)</B
></P
></DIV
></DIV
><DIV
CLASS="sect3"
><HR><H4
CLASS="sect3"
><A
NAME="FilterAlignment"
></A
>Filter Alignments</H4
><P
>&#13;Filtering out the alignments that do not satisfy the specified
conditions. Alignments are filtered from both the alignment and dotplot
displays.  A filter can be applied to one of the following fields:
<P
></P
><UL
><LI
><P
>&#13;<B
CLASS="guimenu"
>Identity</B
>: percentage identity; numeric.
</P
></LI
><LI
><P
>&#13;<B
CLASS="guimenu"
>Score</B
>: similarity score; numeric.
</P
></LI
><LI
><P
>&#13;<B
CLASS="guimenu"
>Best Hit Status</B
>:
one of the following:
3, orthologs (bidirectional best hits, BBH);
2, duplicated only in the sequence-2;
1, duplicated only in the sequence-1;
0, paralogs.
</P
></LI
></UL
>
For specification of the numeric field conditions, see the next subsection
<SPAN
CLASS="QUOTE"
>"<A
HREF="#FilterSegment"
><I
>Filter Genes/Segments</I
></A
>
"</SPAN
>
	  </P
><P
>&#13;Once filtered, a '*' mark is displayed on the left side of the alignment
track (see <A
HREF="#fig_filter_align"
>Figure 11</A
>C).
Additional calls will add filters to the existing one.
For example, users can specify conditions such as "Best Hit Status is BBH
and Identity &#8805; 90%" by two successive calls of this function.
To reset the filters, press the <B
CLASS="guibutton"
>Reset Alignment Filters</B
>
button.
	  </P
><DIV
CLASS="figure"
><A
NAME="fig_filter_align"
></A
><DIV
CLASS="mediaobject"
><P
><IMG
SRC="fig/filter_alignment.gif"></P
></DIV
><P
><B
>Figure 11.  Filtering alignments. The filtering condition window (A), and
alignment displays of before (B) and after (C) filtering (showing only BBH).
</B
></P
></DIV
></DIV
><DIV
CLASS="sect3"
><HR><H4
CLASS="sect3"
><A
NAME="FilterSegment"
></A
>Filter Genes/Segments</H4
><P
>&#13;Filtering out the genes or feature segments that do not satisfy the
specified conditions.  A filter can be applied to any field of any
kind of segment that is currently loaded.
	  </P
><P
>&#13;Kinds of conditions to be applied are different depending on
the types of the fields (numeric and string).
The following conditions are defined:
<P
></P
><DIV
CLASS="variablelist"
><DL
><DT
>Numeric field</DT
><DD
><P
> Available conditions are:
<B
CLASS="guimenu"
>Between</B
> (a &#8804; x &#8804; b),
<B
CLASS="guimenu"
>Except Between</B
>(x &#60; a or b &#60; x),
<B
CLASS="guimenu"
>Less Than</B
> (x &#60; a),
<B
CLASS="guimenu"
>Less Equal</B
> (x &#8804; a),
<B
CLASS="guimenu"
>Equal</B
> (x = a),
<B
CLASS="guimenu"
>Greater Equal</B
> (x &#8805; a), and
<B
CLASS="guimenu"
>Greater Than</B
> (x &#62; a),
where a and/or b are values to be specified.
To specify <B
CLASS="guimenu"
>Between</B
> or <B
CLASS="guimenu"
>Except Between</B
>,
two values should be separated by a comma (e.g. "80,90").
</P
></DD
><DT
>String field</DT
><DD
><P
>&#13;	Available conditions are <B
CLASS="guimenu"
>Regex</B
> (regular expression)
	and  <B
CLASS="guimenu"
>Equal</B
> (exact match).
</P
></DD
></DL
></DIV
>
	  </P
><P
>&#13;Once filtered, a '*' mark is displayed on the left side of the annotation track.
Additional calls will add filters to the existing one.
To reset the filter on the specified segment,
choose the segment type and press the <B
CLASS="guibutton"
>Reset This Filter
</B
> button.
To reset all filters, press the <B
CLASS="guibutton"
>Reset All Filters</B
> button.
	  </P
><DIV
CLASS="figure"
><A
NAME="AEN485"
></A
><DIV
CLASS="mediaobject"
><P
><IMG
SRC="fig/filter_segments.gif"></P
></DIV
><P
><B
>Figure 12.  Filter genes/alignments </B
></P
></DIV
></DIV
><DIV
CLASS="sect3"
><HR><H4
CLASS="sect3"
><A
NAME="DynamicSearch"
></A
>Dynamic Search</H4
><P
>&#13;Search each genome for specific segments or sets of segments
dynamically, by calling a CGI script running on the server.
Results will be displayed in the additional annotation tracks,
or in the outermost pair of tracks (overwriting the existing data)
if the maximum number of segment tracks have already been used.
	  </P
><P
>&#13;Currently, the following programs are available:
	  </P
><P
></P
><UL
><LI
><P
>&#13;<B
CLASS="guimenu"
>Blast</B
>: BLAST similarity search for a nucleotide
[blastn(direct) / tblastx(translation)] or a protein [tblastn] query sequence.
</P
></LI
><LI
><P
>&#13;<B
CLASS="guimenu"
>PatSearch</B
>: Regular expression pattern search. See
the document of the
<A
HREF="http://java.sun.com/j2se/1.4.2/docs/api/java/util/regex/Pattern.html"
TARGET="_top"
>&#13;Java regex library</A
>.
</P
></LI
><LI
><P
>&#13;<B
CLASS="guimenu"
>DirRep</B
>: Direct repeat search.
</P
></LI
><LI
><P
>&#13;<B
CLASS="guimenu"
>SimpleRep</B
>: Simple repeat (or short tandem repeat) search.
</P
></LI
></UL
></DIV
></DIV
><DIV
CLASS="sect2"
><HR><H3
CLASS="sect2"
><A
NAME="View"
></A
>View menu</H3
><DIV
CLASS="sect3"
><H4
CLASS="sect3"
><A
NAME="DataTable"
></A
>Gene/Segment Data Table</H4
><P
>&#13;Show table of genes or specified feature segment on the specified genome.
In the table, each row contains species name, the beginning position,
the ending position, strand direction, assigned color, and the name of
the segment.
Additional fields are added, if any, depending on the segment type.
By clicking a row of this table, the current scope is reset such that
the specified gene or segment is centered.
	  </P
><DIV
CLASS="figure"
><A
NAME="AEN515"
></A
><DIV
CLASS="mediaobject"
><P
><IMG
SRC="fig/genetable.gif"></P
></DIV
><P
><B
>Figure 13.  Gene table </B
></P
></DIV
></DIV
><DIV
CLASS="sect3"
><HR><H4
CLASS="sect3"
><A
NAME="SequenceWindow"
></A
>Sequence Window</H4
><P
>&#13;Show nucleotide sequence of each genome within the current region.
	  </P
></DIV
><DIV
CLASS="sect3"
><HR><H4
CLASS="sect3"
><A
NAME="ExchangeGenomes"
></A
>Exchange Reference Genome</H4
><P
>&#13;Exchange the reference and the target sequences. Both the alignment and
dotplot displays are changed.
	  </P
></DIV
></DIV
></DIV
><DIV
CLASS="sect1"
><HR><H2
CLASS="sect1"
><A
NAME="configAlignViewer"
></A
>Configuring AlignmentViewer</H2
><P
>&#13;Many properties of AlignmentViewer can be configured through property
parameters.  These parameters can be changed on the properties window
(<B
CLASS="guimenu"
>File &#8658; Properties</B
>). Configured properties
are saved in the user home directory.
	</P
><P
>&#13;In the properties window, the properties are categorized into the following
four sections.

<P
></P
><OL
TYPE="1"
><LI
><P
>&#13;<B
CLASS="guimenu"
>Locations</B
>: Locations of data to be loaded.
See <SPAN
CLASS="QUOTE"
>"<A
HREF="#loadData"
><I
>Load data</I
></A
> "</SPAN
>
for loading data.
	<P
></P
><UL
><LI
><P
>&#13;	<B
CLASS="guimenu"
>CGAT_HOME</B
>:
	The server home directory containing the database.
	You can load data from the local file system,
	only when the CGAT database is installed in a directory
	accessible from the AlignmentViewer.
	</P
></LI
><LI
><P
>&#13;	<B
CLASS="guimenu"
>Server URL</B
>:
	Addresses of the CGAT servers. Multiple servers can be specified.
	You can choose which server to use when you load the data.
	</P
></LI
><LI
><P
>&#13;	<B
CLASS="guimenu"
>Proxy URL</B
>:
	Specify if your organization uses a proxy server.
	Please ask your network administrator.
	</P
></LI
></UL
>
</P
></LI
><LI
><P
>&#13;<B
CLASS="guimenu"
>Color</B
>: Color preferences.
You can set two different sets of colors for the dark and light
background color modes.
	<P
></P
><UL
><LI
><P
>&#13;	<B
CLASS="guimenu"
>Background</B
>:
	The background color of both the alignment track and
	the dotplot display.
	</P
></LI
><LI
><P
>&#13;	<B
CLASS="guimenu"
>Viewing Frame</B
>:
	The color of the frame indicating the current scope in the dotplot
	display.
	</P
></LI
><LI
><P
>&#13;	<B
CLASS="guimenu"
>Match/Mismatch/Gap</B
>:
	Colors for displaying nucleotide sequence alignments in the
	annotation track.
	</P
></LI
><LI
><P
>&#13;	<B
CLASS="guimenu"
>Identity</B
>:
	Colors for drawing alignment regions according to the percentage
	identities (the identity color mode). Colors of the three points
	(highest, middle, and lowest percentages) can be specified, and
	the others are interpolated between these colors.
	</P
></LI
><LI
><P
>&#13;	<B
CLASS="guimenu"
>Gene Attribute</B
>:
	Colors for drawing genes in the annotation tracks using gene
	attribute values.  Colors of the highest and lowest values can
	be specified, and the remaining values are interpolated between them.
	</P
></LI
></UL
>
</P
></LI
><LI
><P
>&#13;<B
CLASS="guimenu"
>Panel</B
>: Hight of the tracks and the maximum number of tracks.
	<P
></P
><UL
><LI
><P
>&#13;	<B
CLASS="guimenu"
>Maximum # of Segment Tracks</B
>:
	Maximum number of feature segments to be loaded. Possibly,
	the height of the tracks should also be modified (lowered)
	to display all the information within the screen when you
	want to increase the number of segment tracks.
	</P
></LI
><LI
><P
>&#13;	<B
CLASS="guimenu"
>Height of Alignment/Gene/Segment Area</B
>:
	Hight of the tracks in the alignment display can be set individually
	for the alignment track, the gene annotation tracks and the
	feature segment tracks.
	</P
></LI
></UL
>
</P
></LI
><LI
><P
>&#13;<B
CLASS="guimenu"
>Alignment</B
>: Alignment parameters.
	<P
></P
><UL
><LI
><P
>&#13;	<B
CLASS="guimenu"
>Match/Mismatch/Open Gap/Extension Gap</B
>:
	The scoring system for dynamically calculating alignments
	in the AlignmentViewer program.
	</P
></LI
><LI
><P
>&#13;	<B
CLASS="guimenu"
>Size of Realignment</B
>:
	Maximum length of the displayed region (scope) within which the
	re-alignment of the sequences is allowed by right click on the
	alignment track.
	</P
></LI
></UL
>
</P
></LI
></OL
>
	</P
><DIV
CLASS="figure"
><A
NAME="AEN588"
></A
><DIV
CLASS="mediaobject"
><P
><IMG
SRC="fig/prop_locations.gif"></P
></DIV
><P
><B
>Figure 14. Properties window</B
></P
></DIV
></DIV
></DIV
><DIV
CLASS="chapter"
><HR><H1
><A
NAME="dataserver"
></A
> The CGAT DataServer package</H1
><DIV
CLASS="sect1"
><H2
CLASS="sect1"
><A
NAME="overview"
></A
>Overview of CGAT DataServer</H2
><P
>&#13;The CGAT DataServer package is a collection of data construction scripts and
CGI scripts. The data construction scripts run several programs to generate
comparison data to be visualized by the AlignmentViewer program.
AlignmentViewer can obtain data through the Internet using the CGI
interface implemented in the server package. Alternatively,
AlignmentViewer can also obtain data from the <TT
CLASS="envar"
>$CGAT_HOME</TT
>
directory when the server and client are installed on the same machine.
    </P
><P
>&#13;Data generated by CGAT DataServer can be classified into
the following three classes:
<P
></P
><DIV
CLASS="variablelist"
><DL
><DT
>Alignment</DT
><DD
><P
>&#13;A collection of locally aligned segment pairs between
two genomes, which are displayed on both alignment and dotplot display
panel in AlignmentViewer.
</P
></DD
><DT
>Feature Segments</DT
><DD
><P
>&#13;A collection of feature segments in each genome,
which can be displayed on the feature segment tracks in AlignmentViewer. 
</P
></DD
><DT
>Gene Attribute</DT
><DD
><P
>&#13;A value associated with each gene, which can be used to
assign color to that gene in AlignmentViewer.
</P
></DD
></DL
></DIV
>
    </P
><DIV
CLASS="sect2"
><HR><H3
CLASS="sect2"
><A
NAME="directory"
></A
>Directory structure</H3
><P
>&#13;       <DIV
CLASS="table"
><A
NAME="AEN618"
></A
><P
><B
>Table 1. List of directories on the <TT
CLASS="envar"
>$CGAT_HOME</TT
> directory</B
></P
><TABLE
BORDER="1"
BGCOLOR="#E0E0E0"
CELLSPACING="0"
CELLPADDING="4"
CLASS="CALSTABLE"
><THEAD
><TR
><TH
ALIGN="LEFT"
VALIGN="MIDDLE"
>Directory</TH
><TH
ALIGN="LEFT"
VALIGN="MIDDLE"
>Contents</TH
></TR
></THEAD
><TBODY
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>database</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>The current release of the database, which is accessed by the CGAT CGI scripts</TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>database.work</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>Space for constructing the new database release</TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>work</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>Workspace for database construction where many intermediate files are generated.</TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>perllib</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>Perl libraries.</TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>etc</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>Configuration tables etc.</TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>build</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>Scripts for constructing the database</TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>commands</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>Commands for identifying several feature segments on each genome.</TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>htdocs</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>HTML documents and CGI scripts</TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>client</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>Java source code of the client program (AlignmentViewer)</TD
></TR
></TBODY
></TABLE
></DIV
>
       <DIV
CLASS="table"
><A
NAME="AEN654"
></A
><P
><B
>Table 2. List of directories on the <TT
CLASS="filename"
>database</TT
> directory</B
></P
><TABLE
BORDER="1"
BGCOLOR="#E0E0E0"
CELLSPACING="0"
CELLPADDING="4"
CLASS="CALSTABLE"
><THEAD
><TR
><TH
ALIGN="LEFT"
VALIGN="MIDDLE"
>Directory</TH
><TH
ALIGN="LEFT"
VALIGN="MIDDLE"
>Contents</TH
></TR
></THEAD
><TBODY
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>genomes</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>Genomic sequences</TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>genes</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>Gene information, including position table (tab subdirectory), nucleotide sequences (nt subdirectory), and amino acid sequences (aa subdirectory)</TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>alignment</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> Collection of genome alignments calculated by several programs </TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>segments</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> Collection of feature segments calculated by several programs </TD
></TR
><TR
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
>geneattr</TD
><TD
ALIGN="LEFT"
VALIGN="MIDDLE"
> Collection of gene attribute values calculated by several programs </TD
></TR
></TBODY
></TABLE
></DIV
>
      </P
></DIV
></DIV
><DIV
CLASS="sect1"
><HR><H2
CLASS="sect1"
><A
NAME="AEN678"
></A
>Database construction</H2
><DIV
CLASS="sect2"
><H3
CLASS="sect2"
><A
NAME="databuild"
></A
>Overview of the data construction procedure</H3
><P
>&#13;Essentially, the data construction process consists of the following steps:
     <P
></P
><OL
TYPE="1"
><LI
><P
>&#13;Prepare your genomic sequence data. If your sequences are already
in the <A
HREF="http://mbgd.genome.ad.jp/"
TARGET="_top"
>MBGD</A
>
database, you can skip this step, because
the procedure will automatically try to download missing data from MBGD.
</P
></LI
><LI
><P
>&#13;Edit <TT
CLASS="filename"
><TT
CLASS="envar"
>$CGAT_HOME</TT
>/etc/speclist</TT
>
to specify a set of species to be analyzed and a set of
programs to be executed for each species set.
</P
></LI
><LI
><P
>&#13;Run <B
CLASS="command"
><TT
CLASS="envar"
>$CGAT_HOME</TT
>/build/BuildAll.pl</B
>.
This script executes all procedures for building the database according
to the <TT
CLASS="filename"
><TT
CLASS="envar"
>$CGAT_HOME</TT
>/etc/speclist</TT
> file,
and eventually constructs data on the <TT
CLASS="filename"
>&#13;<TT
CLASS="envar"
>$CGAT_HOME</TT
>/database.work</TT
> directory.
</P
></LI
><LI
><P
>&#13;Run <B
CLASS="command"
><TT
CLASS="envar"
>$CGAT_HOME</TT
>/build/Release.pl</B
> to
release the data from the <TT
CLASS="filename"
><TT
CLASS="envar"
>$CGAT_HOME</TT
>/database.work
</TT
> directory to the <TT
CLASS="filename"
><TT
CLASS="envar"
>$CGAT_HOME</TT
>/database
</TT
> directory.
</P
></LI
></OL
>
    </P
></DIV
><DIV
CLASS="sect2"
><HR><H3
CLASS="sect2"
><A
NAME="prepareData"
></A
>Preparing genomic sequences</H3
><P
>&#13;Before running programs, you must prepare genome sequences data.
There are three possible ways.
<P
></P
><OL
TYPE="1"
><LI
><P
>&#13;If your sequences are already in the
<A
HREF="http://mbgd.genome.ad.jp/"
TARGET="_top"
>MBGD</A
> database,
the simplest way to prepare data is to remember the abbreviated names of those
genomes in the MBGD database (e.g. 'eco' for
<I
CLASS="emphasis"
>Escherichi coli</I
> K12).
and use these names in the configuration table described
in the next section.  The build procedure automatically
try to download missing data from MBGD.
Note that the sequence name specified here is used throughout the system 
(hereafter we call it <TT
CLASS="varname"
>SPNAME</TT
>).
</P
></LI
><LI
><P
>&#13;Alternatively, if you have genomic data in GenBank format, you can use the
<TT
CLASS="filename"
><TT
CLASS="envar"
>$CGAT_HOME</TT
>/build/getDataFromGenBank.pl</TT
>
script to convert it to the CGAT database.
The syntax of the command is
<P
><B
CLASS="command"
>getDataFromGenBank.pl</B
>  <TT
CLASS="replaceable"
><I
>GBK_FILE</I
></TT
>   <TT
CLASS="replaceable"
><I
>SPNAME</I
></TT
> </P
>
For example:
<TABLE
BORDER="0"
BGCOLOR="#E0E0E0"
WIDTH="100%"
><TR
><TD
><PRE
CLASS="screen"
>&#13;getDataFromGenBank.pl NC00913.gbk eco
</PRE
></TD
></TR
></TABLE
>
Here, you can use as <TT
CLASS="varname"
>SPNAME</TT
> any name that is composed
of alphanumeric characters, provided that it is unique.
</P
></LI
><LI
><P
>&#13;Otherwise, you must prepare data by yourself. The data you should prepare
are as follows:
<P
></P
><UL
><LI
><P
>&#13;Choose an appropriate unique name (<TT
CLASS="varname"
>SPNAME</TT
>) that is
composed of alphanumeric characters.
</P
></LI
><LI
><P
>&#13;<TT
CLASS="filename"
>$CGAT_HOME/database/genomes/SPNAME</TT
>:
a genomics sequence in fasta format.
</P
></LI
><LI
><P
>&#13;<TT
CLASS="filename"
>$CGAT_HOME/database/genes/aa/SPNAME</TT
>:
translated sequences of genes in fasta format
(optional; if you want to calculate attribute values associated with
protein sequences).
</P
></LI
><LI
><P
>&#13;<TT
CLASS="filename"
>$CGAT_HOME/database/genes/nt/SPNAME</TT
>:
nucleotide sequences of genes in fasta format
(optional; if you want to calculate attribute values associated with
nucleotide sequences).
</P
></LI
><LI
><P
>&#13;<TT
CLASS="filename"
>$CGAT_HOME/database/genes/tab/SPNAME</TT
>:
a tab-delimited table of genes
containing the following information:
beginning position, ending position,
direction (1/-1), color code (= function category code; optional),
the name of the gene, and the name of the product (optional).
The beginning position must be smaller than the ending position
even for the gene on the reverse chain.
The file must begin with a header line containing a tab-delimited list of
field names as follows,
<TABLE
BORDER="0"
BGCOLOR="#E0E0E0"
WIDTH="100%"
><TR
><TD
><PRE
CLASS="screen"
>&#13;#from	to	dir	color	name	product
190	255	1	1	B0001	thr operon leader peptide
337	2799	1	1	B0002	bifunctional aspartokinase I/homeserine dehydrogenase I
2801	3733	1	1	B0003	homoserine kinase
3734	5020	1	1	B0004	threonine synthase
5234	5530	1	100	B0005	hypothetical protein
</PRE
></TD
></TR
></TABLE
>
By default color code is defined base on the MBGD function category.
Specify '100' if you do not need to assign any color.
You can change the color code by modifying
<TT
CLASS="filename"
>$CGAT_HOME/etc/colorTab/colorTab.gene</TT
>.
</P
></LI
></UL
>
</P
></LI
></OL
>
      </P
></DIV
><DIV
CLASS="sect2"
><HR><H3
CLASS="sect2"
><A
NAME="configBuild"
></A
>Configuring database building procedure</H3
><P
>&#13;Next, you must prepare <TT
CLASS="filename"
>$CGAT_HOME/etc/speclist</TT
> file,
which contains information required for the database building procedure.
The following information should be specified in this file:
<P
></P
><UL
><LI
><P
>&#13;A set of genomic sequences to be compared. If more than two species are
specified, all-against-all pairwise comparison will be performed.
Note that currently CGAT can handle only pairwise comparisons.
</P
></LI
><LI
><P
>&#13;A set of programs to be executed for each sequence set.
</P
></LI
></UL
>
</P
><P
>&#13;The <TT
CLASS="filename"
>speclist</TT
> file consists of macro definition and
dataset definition sections, where macro definition must precede the
dataset definition.  Syntax of the definition of macro variable is
as follows:
<TABLE
BORDER="0"
BGCOLOR="#E0E0E0"
WIDTH="100%"
><TR
><TD
><PRE
CLASS="synopsis"
>&#13;SET <TT
CLASS="varname"
>varname</TT
> = <TT
CLASS="varname"
>value</TT
>
</PRE
></TD
></TR
></TABLE
>
A dollar sign followed by variable name, e.g. $varname, causes a
variable substitution, as usual.
</P
><P
>&#13;The dataset definition section is a tab-delimited table, containing the
following fields: <TT
CLASS="varname"
>SPNAME_LIST</TT
>,
<TT
CLASS="varname"
>PROGRAM_LIST</TT
>,<TT
CLASS="varname"
>FLAG_UPDATE</TT
>,
and <TT
CLASS="varname"
>FLAG_PUBLIC</TT
>.

<TT
CLASS="varname"
>SPNAME_LIST</TT
> is a comma-delimited list specifying
a set of species to be compared.

<TT
CLASS="varname"
>PROGRAM_LIST</TT
> is a space-delimited list specifying
script files to be executed for each species set. The file names
are relative to the <TT
CLASS="filename"
>$CGAT_HOME/build</TT
> directory
and the wildcard characters such as '*' can be used. For example,
<TT
CLASS="filename"
>align/*</TT
> specifies all (executable) files under the
<TT
CLASS="filename"
>$CGAT_HOME/build/align</TT
> directory. 

<TT
CLASS="varname"
>FLAG_UPDATE</TT
> and <TT
CLASS="varname"
>FLAG_PUBLIC</TT
>
are flags specifying the data should be updated and the data should be
open to the public (through the CGI script), respectively.
The value should be 1 (yes) or 0 (no). Default value is 1.
	</P
><P
>&#13;The following is an example of the <TT
CLASS="filename"
>speclist</TT
> file.
The file directs the build script to compare genome sequences
"hpy" and "hpj" by executing all programs for identifying
feature segments in the <TT
CLASS="filename"
>$CGAT_HOME/build/segment</TT
>
directory, followed by all programs for calculating gene attribute values
in the <TT
CLASS="filename"
>$CGAT_HOME/build/geneattr</TT
> directory,
followed by all programs for calculating alignment between the two genomes
n the <TT
CLASS="filename"
>$CGAT_HOME/build/align</TT
> directory.
<TABLE
BORDER="0"
BGCOLOR="#E0E0E0"
WIDTH="100%"
><TR
><TD
><PRE
CLASS="screen"
>&#13;####################
# macro definition
####################
SET AlignAll = align/*
SET SegmentAll = segment/* geneattr/*
####################
# dataset definition
####################
hpy,hpj	$SegmentAll $AlignAll
</PRE
></TD
></TR
></TABLE
>
       </P
></DIV
><DIV
CLASS="sect2"
><HR><H3
CLASS="sect2"
><A
NAME="startBuild"
></A
>Start building</H3
><P
>&#13;Run <B
CLASS="command"
>BuildAll.pl</B
> under the <TT
CLASS="filename"
>$CGAT_HOME/build</TT
>
directory to execute building procedures. This script creates
<TT
CLASS="filename"
>$CGAT_HOME/work</TT
> directory if not exists, and move
to this directory, and then executes programs that have been specified in
the <TT
CLASS="filename"
>$CGAT_HOME/etc/speclist</TT
> file in the previous
section.
<TABLE
BORDER="0"
BGCOLOR="#E0E0E0"
WIDTH="100%"
><TR
><TD
><PRE
CLASS="screen"
>&#13;BuildAll.pl
</PRE
></TD
></TR
></TABLE
>

	</P
></DIV
><DIV
CLASS="sect2"
><HR><H3
CLASS="sect2"
><A
NAME="releaseData"
></A
>Release data</H3
><P
>&#13;Before accessing the created database, you must run
<B
CLASS="command"
>Release.pl</B
> under the <TT
CLASS="filename"
>$CGAT_HOME/build</TT
>
 directory to release the data from the <TT
CLASS="filename"
>$CGAT_HOME/database.work
</TT
> directory to the <TT
CLASS="filename"
>$CGAT_HOME/database
</TT
> directory.
<TABLE
BORDER="0"
BGCOLOR="#E0E0E0"
WIDTH="100%"
><TR
><TD
><PRE
CLASS="screen"
>&#13;Release.pl
</PRE
></TD
></TR
></TABLE
>
    </P
></DIV
></DIV
><DIV
CLASS="sect1"
><HR><H2
CLASS="sect1"
><A
NAME="AEN796"
></A
>Web server set up</H2
><P
>&#13;<TT
CLASS="application"
>AlignmentViewer</TT
> can directly access the resulting
data file locally when the environmental variable <TT
CLASS="envar"
>$CGAT_HOME</TT
>
is correctly set. Therefore you can skip this section if you intend to
use AlignmentViewer only on the machine on which the database was
constructed.
On the other hand, if you want to allow access to the database from
various machines through the Internet, you must set up a web
server on the server machine.
      </P
><P
>&#13;If there is already a HTTP server running on the server, the simplest way
to add the CGAT service to the existing server is to create a symbolic link
to <TT
CLASS="filename"
>$CGAT_HOME/htdocs/CGAT</TT
> on a web site directory.
For example, if you have a <TT
CLASS="filename"
>public_html</TT
> directory
on your home directory as your own homepage, you can add CGAT service
to this page by creating the above symbolic link on this directory:
<TABLE
BORDER="0"
BGCOLOR="#E0E0E0"
WIDTH="100%"
><TR
><TD
><PRE
CLASS="screen"
>&#13;ln -s $CGAT_HOME/htdocs/CGAT ~/public_html
</PRE
></TD
></TR
></TABLE
>
In this case, the URL for accessing the CGAT service is
<TT
CLASS="filename"
>http://YOUR_MACHINE_ADDRESS/~YOUR_USER_NAME/CGAT/</TT
>,
where YOUR_MACHINE_ADDRESS and YOUR_USER_NAME should be replaced with
appropriate names. 
If you are a Mac OSX user, you must have a <TT
CLASS="filename"
>Sites</TT
> 
instead of <TT
CLASS="filename"
>public_html</TT
> directory on your home directory
as your own home page.
      </P
><P
>&#13;Next you must edit the <TT
CLASS="filename"
>httpd.conf</TT
> file to modify
the server setting.

The location of the <TT
CLASS="filename"
>httpd.conf</TT
> file is system
dependent; on Mac OSX, user specific configuration can be described
in a separate file,
<TT
CLASS="filename"
>/private/etc/apache2/users/USERNAME.conf</TT
> (for 10.5) or
<TT
CLASS="filename"
>/etc/httpd/users/USERNAME.conf</TT
> (for 10.4),
where USERNAME should be replaced with your username.

In any way, this step usually requires administrator
privilege. Please ask an administrator of the machine if you do not
have the privilege.
    </P
><P
>&#13;The following is a configuration required for the above example
where the server is located on the public_html directory:
<TABLE
BORDER="0"
BGCOLOR="#E0E0E0"
WIDTH="100%"
><TR
><TD
><PRE
CLASS="screen"
>&#13;
&#60;Directory /PATH_TO_YOUR_HOME_DIRECTORY/public_html&#62;
    Options ExecCGI FollowSymLinks
    AddHandler cgi-script .cgi
    SetEnv CGAT_HOME /PATH_TO_CGAT_HOME
    SetEnv PERLLIB /PATH_TO_CGAT_HOME/perllib/
&#60;/Directory&#62;

</PRE
></TD
></TR
></TABLE
>
where /PATH_TO_YOUR_HOME_DIRECTORY and /PATH_TO_CGAT_HOME should
be replaced with the absolute paths of the respective directories.
    </P
><P
>&#13;Possibly, you may need to load some additional Apache modules.
For example, the above setting contains the SetEnv directive which
requires the mod_env module. If this module is not loaded in the default
setting, you must add the following line to the
<TT
CLASS="filename"
>httpd.conf</TT
> file.
<TABLE
BORDER="0"
BGCOLOR="#E0E0E0"
WIDTH="100%"
><TR
><TD
><PRE
CLASS="screen"
>&#13;LoadModule env_module         MODULE_PATH/mod_env.so
</PRE
></TD
></TR
></TABLE
>
where MODULE_PATH should be replaced with a system dependent module path
(please ask your administrator).
      </P
><P
>&#13;Please remember that you must restart the HTTP server to reflect
the change after you modify the <TT
CLASS="filename"
>httpd.conf</TT
> file.
Typically, this can be done by the command <TT
CLASS="userinput"
><B
>apachectl restart
</B
></TT
>.
Mac OSX users can turn on/off the HTTP server by turning on/off,
respectively, the "Personal Web Sharing" service in the Sharing pane
opened through <B
CLASS="guimenu"
>Apple menu &#8658; System preference &#8658;
Sharing</B
>.
    </P
></DIV
></DIV
></DIV
></BODY
></HTML
>